File: loopx/control_plane/scheduler/state_transition_rules.ts
import type { JsonObject } from "../effect_program.ts";
import { EffectRuntimeRequestError } from "../effect_runtime_errors.ts";
import {
  requireBoolean as requiredBoolean,
  requireInteger as requiredInteger,
  requireJsonObject as requiredObject,
  requireNonEmptyString as requiredString,
} from "../runtime_decode.ts";
import {
  normalizeSchedulerHostUpdateFailures,
  normalizeSchedulerRrule,
  retainedSchedulerHostUpdateFailures,
  rruleForMinutes,
  schedulerRruleIntervalMinutes,
  schedulerTimestampMilliseconds,
} from "./state_store.ts";

export const SCHEDULER_STATE_TRANSITION_REQUEST_SCHEMA =
  "loopx_scheduler_state_transition_request_v0";
export const SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA =
  "loopx_scheduler_state_transition_result_v0";

export const SCHEDULER_CADENCE_TRANSITIONS = [
  "initial",
  "identity_reset",
  "retry_unacknowledged_failure",
  "hold_active_initial",
  "advance_after_interval",
  "hold_until_interval",
] as const;
export type SchedulerCadenceTransition =
  (typeof SCHEDULER_CADENCE_TRANSITIONS)[number];

export const SCHEDULER_HOST_TRANSITIONS = [
  "apply_required",
  "host_match_ack_required",
  "recorded_failure_suppressed",
  "settled",
] as const;
export type SchedulerHostTransition =
  (typeof SCHEDULER_HOST_TRANSITIONS)[number];

export interface SchedulerCadenceTransitionResult extends JsonObject {
  schema_version: typeof SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA;
  operation: "cadence";
  current_index: number;
  state_status: "missing" | "reset_required" | "same_identity";
  transition: SchedulerCadenceTransition;
  current_cadence_acknowledged: boolean;
}

export interface SchedulerHostTransitionResult extends JsonObject {
  schema_version: typeof SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA;
  operation: "host";
  transition: SchedulerHostTransition;
  current_target_has_failure: boolean;
  repeated_failed_pair: boolean;
}

export interface SchedulerBackoffTransitionResult extends JsonObject {
  schema_version: typeof SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA;
  operation: "backoff";
  current_index: number;
  current_interval_minutes: number;
  current_rrule: string;
  last_applied_rrule: string;
  observed_host_rrule: string;
  effective_host_rrule: string;
  state_status: "missing" | "reset_required" | "same_identity";
  cadence_transition: SchedulerCadenceTransition;
  current_cadence_acknowledged: boolean;
  all_host_update_failures: JsonObject[];
  host_update_failures: JsonObject[];
  recorded_host_failure: JsonObject | null;
  current_rrule_already_applied: boolean;
  scheduler_state_acknowledges_current_rrule: boolean;
  host_transition: SchedulerHostTransition;
  current_target_has_failure: boolean;
  repeated_failed_pair: boolean;
}

export interface SchedulerMonitorScheduleResult extends JsonObject {
  schema_version: typeof SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA;
  operation: "monitor_schedule";
  next_due_at: string | null;
  schedule_source: "explicit" | "cadence" | "none";
  cadence_seconds: number | null;
}

export type SchedulerStateTransitionResult =
  | SchedulerCadenceTransitionResult
  | SchedulerHostTransitionResult
  | SchedulerBackoffTransitionResult
  | SchedulerMonitorScheduleResult;

const MONITOR_CADENCE_PATTERN =
  /^\s*(?<count>[1-9][0-9]{0,4})\s*(?<unit>s|sec|secs|second|seconds|m|min|mins|minute|minutes|h|hr|hrs|hour|hours|d|day|days)\s*$/i;

function optionalRequestString(value: unknown, label: string): string {
  if (value === undefined || value === null) return "";
  if (typeof value !== "string") {
    throw new EffectRuntimeRequestError(`${label} must be a string when present`);
  }
  return value.trim();
}

function monitorCadenceSeconds(value: unknown): number | null {
  const cadence = optionalRequestString(value, "monitor cadence");
  if (!cadence) return null;
  const match = MONITOR_CADENCE_PATTERN.exec(cadence);
  if (!match?.groups) return null;
  const count = Number.parseInt(match.groups.count, 10);
  const unit = match.groups.unit.toLowerCase();
  const multiplier = unit.startsWith("s")
    ? 1
    : unit.startsWith("m")
    ? 60
    : unit.startsWith("h")
    ? 60 * 60
    : 24 * 60 * 60;
  return count * multiplier;
}

function monitorScheduleTimestamp(milliseconds: number): string {
  return new Date(milliseconds).toISOString().replace(".000Z", "Z");
}

function evaluateMonitorSchedule(
  request: JsonObject,
): SchedulerMonitorScheduleResult {
  const explicitNextDueAt = optionalRequestString(
    request.explicit_next_due_at,
    "explicit_next_due_at",
  );
  if (explicitNextDueAt) {
    if (schedulerTimestampMilliseconds(explicitNextDueAt) === null) {
      throw new EffectRuntimeRequestError("explicit_next_due_at must be an ISO timestamp");
    }
    return {
      schema_version: SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA,
      operation: "monitor_schedule",
      next_due_at: explicitNextDueAt,
      schedule_source: "explicit",
      cadence_seconds: monitorCadenceSeconds(request.cadence),
    };
  }

  const cadenceSeconds = monitorCadenceSeconds(request.cadence);
  if (cadenceSeconds === null) {
    return {
      schema_version: SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA,
      operation: "monitor_schedule",
      next_due_at: null,
      schedule_source: "none",
      cadence_seconds: null,
    };
  }
  const generatedAt = requiredString(request.generated_at, "generated_at");
  const generatedAtMilliseconds = schedulerTimestampMilliseconds(generatedAt);
  if (generatedAtMilliseconds === null) {
    throw new EffectRuntimeRequestError("generated_at must be an ISO timestamp");
  }
  return {
    schema_version: SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA,
    operation: "monitor_schedule",
    next_due_at: monitorScheduleTimestamp(
      generatedAtMilliseconds + cadenceSeconds * 1_000,
    ),
    schedule_source: "cadence",
    cadence_seconds: cadenceSeconds,
  };
}

function requiredPositiveIntegerList(value: unknown, label: string): number[] {
  if (
    !Array.isArray(value) ||
    value.length === 0 ||
    value.some((item) =>
      typeof item !== "number" || !Number.isInteger(item) || item <= 0
    )
  ) {
    throw new EffectRuntimeRequestError(`${label} must be a non-empty array of positive integers`);
  }
  return [...value] as number[];
}

function requestObject(value: unknown): JsonObject {
  const request = requiredObject(value, "scheduler.state_transition params");
  if (request.schema_version !== SCHEDULER_STATE_TRANSITION_REQUEST_SCHEMA) {
    throw new EffectRuntimeRequestError("Scheduler state transition request schema mismatch");
  }
  return request;
}

function evaluateCadence(request: JsonObject): SchedulerCadenceTransitionResult {
  const progressionSize = requiredInteger(
    request.progression_size,
    "progression_size",
  );
  if (progressionSize < 1) {
    throw new EffectRuntimeRequestError("scheduler cadence progression must not be empty");
  }
  if (!requiredBoolean(request.state_present, "state_present")) {
    return {
      schema_version: SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA,
      operation: "cadence",
      current_index: 0,
      state_status: "missing",
      transition: "initial",
      current_cadence_acknowledged: false,
    };
  }
  if (!requiredBoolean(request.identity_matches, "identity_matches")) {
    return {
      schema_version: SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA,
      operation: "cadence",
      current_index: 0,
      state_status: "reset_required",
      transition: "identity_reset",
      current_cadence_acknowledged: false,
    };
  }

  const appliedIndex = requiredInteger(request.applied_index, "applied_index");
  const acknowledged = requiredBoolean(
    request.current_cadence_acknowledged,
    "current_cadence_acknowledged",
  );
  let nextIndex: number;
  let transition: SchedulerCadenceTransition;
  if (requiredBoolean(request.has_host_update_failures, "has_host_update_failures") &&
    !acknowledged) {
    nextIndex = appliedIndex;
    transition = "retry_unacknowledged_failure";
  } else if (!requiredBoolean(request.advance_same_identity, "advance_same_identity")) {
    nextIndex = 0;
    transition = "hold_active_initial";
  } else if (requiredBoolean(request.applied_interval_elapsed, "applied_interval_elapsed")) {
    nextIndex = appliedIndex + 1;
    transition = "advance_after_interval";
  } else {
    nextIndex = appliedIndex;
    transition = "hold_until_interval";
  }
  return {
    schema_version: SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA,
    operation: "cadence",
    current_index: Math.min(Math.max(nextIndex, 0), progressionSize - 1),
    state_status: "same_identity",
    transition,
    current_cadence_acknowledged: acknowledged,
  };
}

function evaluateHost(request: JsonObject): SchedulerHostTransitionResult {
  const stateStatus = requiredString(request.state_status, "state_status");
  const currentTargetHasFailure = requiredBoolean(
    request.current_target_has_failure,
    "current_target_has_failure",
  );
  const repeatedFailedPair = requiredBoolean(
    request.repeated_failed_pair,
    "repeated_failed_pair",
  );
  let transition: SchedulerHostTransition;
  if (
    requiredBoolean(request.observed_host_rrule_present, "observed_host_rrule_present") &&
    requiredBoolean(request.current_rrule_already_applied, "current_rrule_already_applied") &&
    (!requiredBoolean(
      request.scheduler_state_acknowledges_current_rrule,
      "scheduler_state_acknowledges_current_rrule",
    ) || stateStatus !== "same_identity" || currentTargetHasFailure)
  ) {
    transition = "host_match_ack_required";
  } else if (
    requiredBoolean(request.current_rrule_already_applied, "current_rrule_already_applied") &&
    stateStatus === "same_identity"
  ) {
    transition = "settled";
  } else if (repeatedFailedPair) {
    transition = "recorded_failure_suppressed";
  } else {
    transition = "apply_required";
  }
  return {
    schema_version: SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA,
    operation: "host",
    transition,
    current_target_has_failure: currentTargetHasFailure,
    repeated_failed_pair: repeatedFailedPair,
  };
}

function schedulerAppliedIndex(
  progressionMinutes: readonly number[],
  schedulerState: JsonObject,
): number {
  const value = schedulerState.progression_index;
  return typeof value === "number" &&
      Number.isInteger(value) &&
      value >= 0 &&
      value < progressionMinutes.length
    ? value
    : -1;
}

function schedulerProgressionIntervalElapsed(
  schedulerState: JsonObject,
  currentTime: string,
): boolean {
  const updatedAt = schedulerTimestampMilliseconds(schedulerState.updated_at);
  const current = schedulerTimestampMilliseconds(currentTime);
  const appliedInterval = schedulerRruleIntervalMinutes(
    schedulerState.last_applied_rrule,
  );
  if (updatedAt === null || current === null || appliedInterval === null) {
    return true;
  }
  return current >= updatedAt + appliedInterval * 60_000;
}

function monitorRruleAppliedWithinStaleTolerance(
  cadenceClass: string,
  lastAppliedRrule: string,
  currentRrule: string,
  toleranceMinutes: number,
): boolean {
  if (cadenceClass !== "monitor_wait") return false;
  const appliedMinutes = schedulerRruleIntervalMinutes(lastAppliedRrule);
  const currentMinutes = schedulerRruleIntervalMinutes(currentRrule);
  return appliedMinutes !== null &&
    currentMinutes !== null &&
    appliedMinutes >= currentMinutes &&
    appliedMinutes - currentMinutes <= toleranceMinutes;
}

function evaluateBackoff(request: JsonObject): SchedulerBackoffTransitionResult {
  const progressionMinutes = requiredPositiveIntegerList(
    request.progression_minutes,
    "progression_minutes",
  );
  const schedulerState = requiredObject(
    request.scheduler_state,
    "scheduler_state",
  );
  const resetToken = requiredString(request.reset_token, "reset_token");
  const identitySignature = requiredString(
    request.identity_signature,
    "identity_signature",
  );
  const currentTime = requiredString(request.current_time, "current_time");
  const cadenceClass = requiredString(request.cadence_class, "cadence_class");
  const staleToleranceMinutes = requiredInteger(
    request.stale_tolerance_minutes,
    "stale_tolerance_minutes",
  );
  if (staleToleranceMinutes < 0) {
    throw new EffectRuntimeRequestError("stale_tolerance_minutes must not be negative");
  }

  const allHostUpdateFailures = retainedSchedulerHostUpdateFailures(
    normalizeSchedulerHostUpdateFailures(
      schedulerState.host_update_failures,
      schedulerState.host_update_failure,
    ),
    currentTime,
  );
  const appliedIndex = schedulerAppliedIndex(
    progressionMinutes,
    schedulerState,
  );
  const appliedTargetRrule = appliedIndex >= 0
    ? rruleForMinutes(progressionMinutes[appliedIndex])
    : "";
  const lastAppliedRrule = typeof schedulerState.last_applied_rrule === "string"
    ? schedulerState.last_applied_rrule.trim()
    : "";
  const currentCadenceAcknowledged = Boolean(appliedTargetRrule) &&
    normalizeSchedulerRrule(lastAppliedRrule) === appliedTargetRrule;
  const cadence = evaluateCadence({
    progression_size: progressionMinutes.length,
    state_present: Object.keys(schedulerState).length > 0,
    identity_matches: Object.keys(schedulerState).length > 0 &&
      schedulerState.reset_token === resetToken &&
      schedulerState.identity_signature === identitySignature,
    applied_index: appliedIndex,
    current_cadence_acknowledged: currentCadenceAcknowledged,
    advance_same_identity: requiredBoolean(
      request.advance_same_identity,
      "advance_same_identity",
    ),
    applied_interval_elapsed: schedulerProgressionIntervalElapsed(
      schedulerState,
      currentTime,
    ),
    has_host_update_failures: allHostUpdateFailures.length > 0,
  });
  const currentIndex = cadence.current_index;
  const currentInterval = progressionMinutes[currentIndex];
  const currentRrule = rruleForMinutes(currentInterval);
  const observedHostRrule = normalizeSchedulerRrule(
    request.observed_host_rrule,
  );
  const effectiveHostRrule = observedHostRrule || lastAppliedRrule;
  const hostUpdateFailures = retainedSchedulerHostUpdateFailures(
    allHostUpdateFailures,
    currentTime,
    effectiveHostRrule,
  );
  const recordedHostFailure = [...hostUpdateFailures].reverse().find((failure) =>
    normalizeSchedulerRrule(failure.target_rrule) === currentRrule
  ) ?? hostUpdateFailures.at(-1) ?? null;
  let currentRruleAlreadyApplied = effectiveHostRrule === currentRrule;
  if (
    !currentRruleAlreadyApplied &&
    !observedHostRrule &&
    cadence.state_status === "same_identity"
  ) {
    currentRruleAlreadyApplied = monitorRruleAppliedWithinStaleTolerance(
      cadenceClass,
      effectiveHostRrule,
      currentRrule,
      staleToleranceMinutes,
    );
  }
  const schedulerStateAcknowledgesCurrentRrule =
    cadence.state_status === "same_identity" &&
    normalizeSchedulerRrule(lastAppliedRrule) === currentRrule;
  const currentTargetHasFailure = allHostUpdateFailures.some((failure) =>
    normalizeSchedulerRrule(failure.target_rrule) === currentRrule
  );
  const failedTargetRrule = normalizeSchedulerRrule(
    recordedHostFailure?.target_rrule,
  );
  const failedObservedHostRrule = normalizeSchedulerRrule(
    recordedHostFailure?.observed_host_rrule,
  );
  const repeatedFailedPair = Boolean(
    failedTargetRrule === currentRrule &&
      failedObservedHostRrule === effectiveHostRrule,
  );
  const host = evaluateHost({
    state_status: cadence.state_status,
    observed_host_rrule_present: Boolean(observedHostRrule),
    current_rrule_already_applied: currentRruleAlreadyApplied,
    scheduler_state_acknowledges_current_rrule:
      schedulerStateAcknowledgesCurrentRrule,
    current_target_has_failure: currentTargetHasFailure,
    repeated_failed_pair: repeatedFailedPair,
  });
  return {
    schema_version: SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA,
    operation: "backoff",
    current_index: currentIndex,
    current_interval_minutes: currentInterval,
    current_rrule: currentRrule,
    last_applied_rrule: lastAppliedRrule,
    observed_host_rrule: observedHostRrule,
    effective_host_rrule: effectiveHostRrule,
    state_status: cadence.state_status,
    cadence_transition: cadence.transition,
    current_cadence_acknowledged: currentCadenceAcknowledged,
    all_host_update_failures: allHostUpdateFailures,
    host_update_failures: hostUpdateFailures,
    recorded_host_failure: recordedHostFailure,
    current_rrule_already_applied: currentRruleAlreadyApplied,
    scheduler_state_acknowledges_current_rrule:
      schedulerStateAcknowledgesCurrentRrule,
    host_transition: host.transition,
    current_target_has_failure: currentTargetHasFailure,
    repeated_failed_pair: repeatedFailedPair,
  };
}

export function evaluateSchedulerStateTransition(
  value: unknown,
): SchedulerStateTransitionResult {
  const request = requestObject(value);
  if (request.operation === "cadence") return evaluateCadence(request);
  if (request.operation === "host") return evaluateHost(request);
  if (request.operation === "backoff") return evaluateBackoff(request);
  if (request.operation === "monitor_schedule") {
    return evaluateMonitorSchedule(request);
  }
  throw new EffectRuntimeRequestError("Scheduler state transition operation is unsupported");
}

File: loopx/control_plane/scheduler/state_transition_rules.py
from __future__ import annotations

from collections.abc import Collection, Mapping, Sequence
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Any

from ..effect_runtime import EffectRuntimeRejected, effect_runtime_result
from .state import normalize_scheduler_rrule, rrule_for_minutes


SCHEDULER_STATE_TRANSITION_REQUEST_SCHEMA = (
    "loopx_scheduler_state_transition_request_v0"
)
SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA = (
    "loopx_scheduler_state_transition_result_v0"
)


class SchedulerCadenceTransition(str, Enum):
    INITIAL = "initial"
    IDENTITY_RESET = "identity_reset"
    RETRY_UNACKNOWLEDGED_FAILURE = "retry_unacknowledged_failure"
    HOLD_ACTIVE_INITIAL = "hold_active_initial"
    ADVANCE_AFTER_INTERVAL = "advance_after_interval"
    HOLD_UNTIL_INTERVAL = "hold_until_interval"


@dataclass(frozen=True)
class SchedulerCadenceDecision:
    current_index: int
    state_status: str
    transition: SchedulerCadenceTransition
    current_cadence_acknowledged: bool


@dataclass(frozen=True)
class MonitorScheduleProjection:
    next_due_at: str | None
    schedule_source: str
    cadence_seconds: int | None


def project_monitor_todo_schedule(
    *,
    generated_at: str,
    cadence: Any = None,
    explicit_next_due_at: Any = None,
) -> MonitorScheduleProjection:
    """Project monitor schedule facts through the TypeScript scheduler owner."""

    result = _runtime_result(
        {
            "schema_version": SCHEDULER_STATE_TRANSITION_REQUEST_SCHEMA,
            "operation": "monitor_schedule",
            "generated_at": generated_at,
            "cadence": None if cadence is None else str(cadence),
            "explicit_next_due_at": (
                None
                if explicit_next_due_at is None
                else str(explicit_next_due_at)
            ),
        }
    )
    next_due_at = result.get("next_due_at")
    schedule_source = result.get("schedule_source")
    cadence_seconds = result.get("cadence_seconds")
    if (
        result.get("operation") != "monitor_schedule"
        or (next_due_at is not None and not isinstance(next_due_at, str))
        or schedule_source not in {"explicit", "cadence", "none"}
        or (
            cadence_seconds is not None
            and (
                isinstance(cadence_seconds, bool)
                or not isinstance(cadence_seconds, int)
                or cadence_seconds <= 0
            )
        )
    ):
        raise RuntimeError("TypeScript monitor schedule result shape mismatch")
    return MonitorScheduleProjection(
        next_due_at=next_due_at,
        schedule_source=str(schedule_source),
        cadence_seconds=cadence_seconds,
    )


def _runtime_result(params: dict[str, Any]) -> Mapping[str, Any]:
    try:
        result = effect_runtime_result("scheduler.state_transition.evaluate", params)
    except EffectRuntimeRejected as exc:
        raise ValueError(str(exc)) from None
    if not isinstance(result, Mapping):
        raise RuntimeError("TypeScript scheduler transition result must be an object")
    if result.get("schema_version") != SCHEDULER_STATE_TRANSITION_RESULT_SCHEMA:
        raise RuntimeError("TypeScript scheduler transition result shape mismatch")
    return result


def _scheduler_applied_index(
    progression_minutes: Sequence[int], scheduler_state: Mapping[str, Any]
) -> int:
    raw_index = scheduler_state.get("progression_index")
    if raw_index is None:
        return -1
    try:
        applied_index = int(raw_index)
    except (TypeError, ValueError):
        return -1
    return applied_index if 0 <= applied_index < len(progression_minutes) else -1


def decide_scheduler_cadence_transition(
    progression_minutes: Sequence[int],
    *,
    scheduler_state: Mapping[str, Any],
    reset_token: str,
    identity_signature: str,
    advance_same_identity: bool,
    applied_interval_elapsed: bool,
    has_host_update_failures: bool,
) -> SchedulerCadenceDecision:
    """Choose the cadence index through the TypeScript transition kernel."""

    if not progression_minutes:
        raise ValueError("scheduler cadence progression must not be empty")
    identity_matches = bool(scheduler_state) and (
        scheduler_state.get("reset_token") == reset_token
        and scheduler_state.get("identity_signature") == identity_signature
    )
    applied_index = _scheduler_applied_index(progression_minutes, scheduler_state)
    applied_target_rrule = (
        rrule_for_minutes(progression_minutes[applied_index])
        if 0 <= applied_index < len(progression_minutes)
        else ""
    )
    current_cadence_acknowledged = (
        bool(applied_target_rrule)
        and normalize_scheduler_rrule(scheduler_state.get("last_applied_rrule"))
        == applied_target_rrule
    )
    result = _runtime_result(
        {
            "schema_version": SCHEDULER_STATE_TRANSITION_REQUEST_SCHEMA,
            "operation": "cadence",
            "progression_size": len(progression_minutes),
            "state_present": bool(scheduler_state),
            "identity_matches": identity_matches,
            "applied_index": applied_index,
            "current_cadence_acknowledged": current_cadence_acknowledged,
            "advance_same_identity": advance_same_identity,
            "applied_interval_elapsed": applied_interval_elapsed,
            "has_host_update_failures": has_host_update_failures,
        }
    )
    transition = result.get("transition")
    current_index = result.get("current_index")
    state_status = result.get("state_status")
    acknowledged = result.get("current_cadence_acknowledged")
    if (
        result.get("operation") != "cadence"
        or isinstance(current_index, bool)
        or not isinstance(current_index, int)
        or not isinstance(state_status, str)
        or transition not in {item.value for item in SchedulerCadenceTransition}
        or not isinstance(acknowledged, bool)
    ):
        raise RuntimeError("TypeScript scheduler cadence result shape mismatch")
    return SchedulerCadenceDecision(
        current_index=current_index,
        state_status=state_status,
        transition=SchedulerCadenceTransition(str(transition)),
        current_cadence_acknowledged=acknowledged,
    )


class SchedulerHostTransition(str, Enum):
    APPLY_REQUIRED = "apply_required"
    HOST_MATCH_ACK_REQUIRED = "host_match_ack_required"
    RECORDED_FAILURE_SUPPRESSED = "recorded_failure_suppressed"
    SETTLED = "settled"


@dataclass(frozen=True)
class SchedulerHostDecision:
    transition: SchedulerHostTransition
    current_target_has_failure: bool
    repeated_failed_pair: bool

    @property
    def apply_needed(self) -> bool:
        return self.transition == SchedulerHostTransition.APPLY_REQUIRED

    @property
    def host_match_ack_needed(self) -> bool:
        return self.transition == SchedulerHostTransition.HOST_MATCH_ACK_REQUIRED

    @property
    def host_failure_suppressed(self) -> bool:
        return self.transition == SchedulerHostTransition.RECORDED_FAILURE_SUPPRESSED

    @property
    def ack_needed(self) -> bool:
        return self.transition in {
            SchedulerHostTransition.APPLY_REQUIRED,
            SchedulerHostTransition.HOST_MATCH_ACK_REQUIRED,
        }


@dataclass(frozen=True)
class SchedulerBackoffDecision:
    cadence: SchedulerCadenceDecision
    host: SchedulerHostDecision
    current_interval_minutes: int
    current_rrule: str
    last_applied_rrule: str
    observed_host_rrule: str
    effective_host_rrule: str
    all_host_update_failures: tuple[dict[str, Any], ...]
    host_update_failures: tuple[dict[str, Any], ...]
    recorded_host_failure: dict[str, Any] | None
    current_rrule_already_applied: bool
    scheduler_state_acknowledges_current_rrule: bool


def _object_tuple(value: Any, *, label: str) -> tuple[dict[str, Any], ...]:
    if not isinstance(value, list) or any(not isinstance(item, dict) for item in value):
        raise RuntimeError(f"TypeScript {label} must be an object list")
    return tuple(dict(item) for item in value)


def decide_scheduler_backoff_state(
    progression_minutes: Sequence[int],
    *,
    scheduler_state: Mapping[str, Any],
    reset_token: str,
    identity_signature: str,
    advance_same_identity: bool,
    current_time: datetime | str,
    observed_host_rrule: Any,
    cadence_class: str,
    stale_tolerance_minutes: int,
) -> SchedulerBackoffDecision:
    """Evaluate one complete scheduler backoff slice in the TS owner."""

    result = _runtime_result(
        {
            "schema_version": SCHEDULER_STATE_TRANSITION_REQUEST_SCHEMA,
            "operation": "backoff",
            "progression_minutes": list(progression_minutes),
            "scheduler_state": dict(scheduler_state),
            "reset_token": reset_token,
            "identity_signature": identity_signature,
            "advance_same_identity": advance_same_identity,
            "current_time": (
                current_time.isoformat()
                if isinstance(current_time, datetime)
                else current_time
            ),
            "observed_host_rrule": observed_host_rrule,
            "cadence_class": cadence_class,
            "stale_tolerance_minutes": stale_tolerance_minutes,
        }
    )
    current_index = result.get("current_index")
    current_interval = result.get("current_interval_minutes")
    state_status = result.get("state_status")
    cadence_transition = result.get("cadence_transition")
    cadence_acknowledged = result.get("current_cadence_acknowledged")
    host_transition = result.get("host_transition")
    target_has_failure = result.get("current_target_has_failure")
    repeated_failed_pair = result.get("repeated_failed_pair")
    current_rrule_applied = result.get("current_rrule_already_applied")
    state_acknowledges_rrule = result.get(
        "scheduler_state_acknowledges_current_rrule"
    )
    current_rrule = result.get("current_rrule")
    last_applied_rrule = result.get("last_applied_rrule")
    observed_result_rrule = result.get("observed_host_rrule")
    effective_host_rrule = result.get("effective_host_rrule")
    recorded_failure = result.get("recorded_host_failure")
    if (
        result.get("operation") != "backoff"
        or isinstance(current_index, bool)
        or not isinstance(current_index, int)
        or isinstance(current_interval, bool)
        or not isinstance(current_interval, int)
        or not isinstance(state_status, str)
        or cadence_transition not in {item.value for item in SchedulerCadenceTransition}
        or not isinstance(cadence_acknowledged, bool)
        or host_transition not in {item.value for item in SchedulerHostTransition}
        or not isinstance(target_has_failure, bool)
        or not isinstance(repeated_failed_pair, bool)
        or not isinstance(current_rrule_applied, bool)
        or not isinstance(state_acknowledges_rrule, bool)
        or not isinstance(current_rrule, str)
        or not isinstance(last_applied_rrule, str)
        or not isinstance(observed_result_rrule, str)
        or not isinstance(effective_host_rrule, str)
        or (recorded_failure is not None and not isinstance(recorded_failure, dict))
    ):
        raise RuntimeError("TypeScript scheduler backoff result shape mismatch")
    cadence = SchedulerCadenceDecision(
        current_index=current_index,
        state_status=state_status,
        transition=SchedulerCadenceTransition(str(cadence_transition)),
        current_cadence_acknowledged=cadence_acknowledged,
    )
    host = SchedulerHostDecision(
        transition=SchedulerHostTransition(str(host_transition)),
        current_target_has_failure=target_has_failure,
        repeated_failed_pair=repeated_failed_pair,
    )
    return SchedulerBackoffDecision(
        cadence=cadence,
        host=host,
        current_interval_minutes=current_interval,
        current_rrule=current_rrule,
        last_applied_rrule=last_applied_rrule,
        observed_host_rrule=observed_result_rrule,
        effective_host_rrule=effective_host_rrule,
        all_host_update_failures=_object_tuple(
            result.get("all_host_update_failures"),
            label="scheduler all-host failure cache",
        ),
        host_update_failures=_object_tuple(
            result.get("host_update_failures"),
            label="scheduler host failure cache",
        ),
        recorded_host_failure=(
            dict(recorded_failure) if isinstance(recorded_failure, dict) else None
        ),
        current_rrule_already_applied=current_rrule_applied,
        scheduler_state_acknowledges_current_rrule=state_acknowledges_rrule,
    )


def decide_scheduler_host_transition(
    *,
    state_status: str,
    observed_host_rrule: str,
    effective_host_rrule: str,
    current_rrule: str,
    current_rrule_already_applied: bool,
    scheduler_state_acknowledges_current_rrule: bool,
    all_host_update_failures: Collection[Mapping[str, Any]],
    recorded_host_failure: Mapping[str, Any] | None,
) -> SchedulerHostDecision:
    """Classify the host action through the TypeScript transition kernel."""

    current_target_has_failure = any(
        normalize_scheduler_rrule(failure.get("target_rrule")) == current_rrule
        for failure in all_host_update_failures
    )
    failed_target_rrule = normalize_scheduler_rrule(
        (recorded_host_failure or {}).get("target_rrule")
    )
    failed_observed_host_rrule = normalize_scheduler_rrule(
        (recorded_host_failure or {}).get("observed_host_rrule")
    )
    repeated_failed_pair = bool(
        failed_target_rrule == current_rrule
        and failed_observed_host_rrule == effective_host_rrule
    )

    result = _runtime_result(
        {
            "schema_version": SCHEDULER_STATE_TRANSITION_REQUEST_SCHEMA,
            "operation": "host",
            "state_status": state_status,
            "observed_host_rrule_present": bool(observed_host_rrule),
            "current_rrule_already_applied": current_rrule_already_applied,
            "scheduler_state_acknowledges_current_rrule": (
                scheduler_state_acknowledges_current_rrule
            ),
            "current_target_has_failure": current_target_has_failure,
            "repeated_failed_pair": repeated_failed_pair,
        }
    )
    transition = result.get("transition")
    result_has_failure = result.get("current_target_has_failure")
    result_repeated_pair = result.get("repeated_failed_pair")
    if (
        result.get("operation") != "host"
        or transition not in {item.value for item in SchedulerHostTransition}
        or not isinstance(result_has_failure, bool)
        or not isinstance(result_repeated_pair, bool)
    ):
        raise RuntimeError("TypeScript scheduler host result shape mismatch")
    return SchedulerHostDecision(
        transition=SchedulerHostTransition(str(transition)),
        current_target_has_failure=result_has_failure,
        repeated_failed_pair=result_repeated_pair,
    )
